src/rust/engine/src/externs/interface.rs

```diff
@@ -830,6 +840,40 @@ fn nailgun_server_await_shutdown(py: Python, nailgun_server_ptr: PyNailgunServer
   })
 }
 
+fn strongly_connected_components(
+  py: Python,
+  adjacency_lists: Vec<(PyObject, Vec<PyObject>)>,
+) -> CPyResult<Vec<Vec<PyObject>>> {
```

Comment:
It'd be great to do this w/ PyO3, but I see it depends on `key_for` which hasn't been ported yet. I can port this in a followup - it's more important we land this to unblock Patrick.

Reply:
Yea. You can see in an earlier commit where I was attempting to do that too: the issue was that without hash/eq, I'd need to use PyDict rather than the rust native hashmap (not much of a hardship...maybe even more performant, but awkward).

Reply:
What do you mean without hash/eq?

Reply:
`PyObject` itself does not have Eq/Hash implementations, so you need wrapper types around it to add those (we use `Value` and `Key` to do that).